src/Facet/FacetContainer.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Wikibase\DataModel\Facet;
+
+/**
+ * Interface for attaching and accessing facets (aka "roles objects" or "extensions").
+ *
+ * This interface provides the basis of an implementation of the Role Object Pattern, see
+ * <http://c2.com/cgi/wiki?RoleObjectPattern>.
+ *
+ * @since 5.0
+ *
+ * @licence GNU GPL v2+
+ * @author Daniel Kinzler
+ */
+interface FacetContainer {
+
+	/**
+	 * Determines whether the given facet is supported.
+	 *
+	 * @param string $name
+	 *
+	 * @return boolean
+	 */
+	public function hasFacet( $name );
+
+	/**
+	 * Returns a list of supported facet names.
+	 *
+	 * @return string[]
+	 */
+	public function listFacets();
+
+	/**
+	 * Returns a facet object.
+	 *
+	 * @param string $name
+	 * @param string|null $type The desired type
+	 *
+	 * @throws NoSuchFacetException if no facet object is found for $name
+	 * @throws MismatchingFacetException if the facet object is not compatible with $type
+	 * @return object
+	 */
+	public function getFacet( $name, $type = null );
+
+	/**
+	 * Adds a facet object. Any facet object previously registered under the same name
```

Comment:
I wonder why this interface does have a setter/adder at all? I expected the knowledge which facets are supported, how they are build and so on to be an implementation detail of the class that implements the interface. But with this public (!) setter everybody is free to add whatever object they want and call it a "facet". Why is this needed? Can't this be protected or omitted?